# Terragrunt v0.19.30 breaks get_aws_account_id() functionality

**sc250024** commented *Oct 8, 2019*

## Description

Greetings,

Upgrading to Terragrunt v0.19.30 breaks using the `get_aws_account_id()` function to assume a role for an S3 backend configuration.

Consider the following `terragrunt.hcl` file at the root of my Terraform repository:

<details><summary><b>terragrunt.hcl</b></summary>
<p>

```hcl
remote_state {
  backend = "s3"

  config = {
    bucket         = "some-terraform-state-bucket"
    dynamodb_table = "some-terraform-lock-table"
    encrypt        = true
    key            = "${path_relative_to_include()}/terraform.tfstate"
    region         = "eu-west-1"
    role_arn       = "arn:aws:iam::${get_env("TG_AWS_ACCT", "${get_aws_account_id()}")}:role/TerragruntAdministrator"
  }
}

terraform {
  extra_arguments "conditional_vars" {
    commands = get_terraform_commands_that_need_vars()

    required_var_files = [
      "${get_parent_terragrunt_dir()}/global.tfvars"
    ]

    optional_var_files = [
      "${find_in_parent_folders("environment.tfvars", "ignore")}",
      "${find_in_parent_folders("account.tfvars", "ignore")}",
      "${find_in_parent_folders("region.tfvars", "ignore")}",
      "${find_in_parent_folders("stage.tfvars", "ignore")}"
    ]
  }

  extra_arguments "disable_input" {
    commands  = get_terraform_commands_that_need_input()
    arguments = ["-input=false"]
  }

  extra_arguments "retry_lock" {
    commands  = get_terraform_commands_that_need_locking()
    arguments = ["-lock-timeout=5m"]
  }
}
```

</p>
</details>

Running the command `terragrunt plan` using version **v0.19.29** works as expected:

<details><summary><b>Terragrunt v0.19.29 plan output</b></summary>
<p>

```shell
$ ./terragrunt --version
terragrunt version v0.19.29

$ TERRAGRUNT_DEBUG='true' ./terragrunt plan
[terragrunt] 2019/10/08 10:16:49 Reading Terragrunt config file at /path/to/my/terragrunt/module/terragrunt.hcl
[terragrunt] [/path/to/my/terragrunt/module] 2019/10/08 10:16:49 Running command: terraform --version
[terragrunt] 2019/10/08 10:16:50 Running command: terraform plan -var-file=/path/to/my/terragrunt/global.tfvars -var-file=../../../../environment.tfvars -var-file=../../../account.tfvars -var-file=../../region.tfvars -var-file=../stage.tfvars -input=false -lock-timeout=5m
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_caller_identity.current: Refreshing state...
data.terraform_remote_state.organization: Refreshing state...
module.vpc_label.data.aws_region.current: Refreshing state...
module.vpc.module.this_vpc.data.aws_vpc_endpoint_service.dynamodb[0]: Refreshing state...
module.vpc.module.this_vpc.data.aws_vpc_endpoint_service.s3[0]: Refreshing state...
module.vpc.data.aws_iam_policy_document.this_vpc_flow_log_role_policy[0]: Refreshing state...
module.vpc.data.aws_availability_zones.available: Refreshing state...
module.vpc.data.aws_iam_policy_document.this_vpc_flow_log_role_assume_policy[0]: Refreshing state...
module.vpc_label.module.this_label.data.null_data_source.tags_as_list_of_maps[4]: Refreshing state...
module.vpc_label.module.this_label.data.null_data_source.tags_as_list_of_maps[0]: Refreshing state...
module.vpc_label.module.this_label.data.null_data_source.tags_as_list_of_maps[1]: Refreshing state...
module.vpc_label.module.this_label.data.null_data_source.tags_as_list_of_maps[2]: Refreshing state...
module.vpc_label.module.this_label.data.null_data_source.tags_as_list_of_maps[5]: Refreshing state...
module.vpc_label.module.this_label.data.null_data_source.tags_as_list_of_maps[3]: Refreshing state...
module.vpc.aws_cloudwatch_log_group.this_vpc[0]: Refreshing state... [id=/aws/vpc/my-vpc/flow-logs]
module.vpc.aws_key_pair.this_vpc: Refreshing state... [id=my-vpc-default-key]
module.vpc.aws_iam_role.this_vpc[0]: Refreshing state... [id=my-vpc-flow-log-role]
module.vpc.module.this_vpc.aws_vpc.this[0]: Refreshing state... [id=vpc-07cec2b6c692203c2]
module.vpc.aws_eip.this_vpc[2]: Refreshing state... [id=eipalloc-0b6f305ad47bf3532]
module.vpc.aws_eip.this_vpc[1]: Refreshing state... [id=eipalloc-0e22269636ad45c70]
module.vpc.aws_eip.this_vpc[0]: Refreshing state... [id=eipalloc-00e7564a08b685ccf]
module.vpc.aws_iam_role_policy.this_vpc[0]: Refreshing state... [id=my-vpc-flow-log-role:my-vpc-flow-log-policy]
module.vpc.data.aws_security_group.default: Refreshing state...
module.vpc.aws_flow_log.this_vpc[0]: Refreshing state... [id=fl-0de32fa0a3ca7c716]
module.vpc.module.this_vpc.aws_subnet.database[0]: Refreshing state... [id=subnet-0f8f0bd12296b1a36]
module.vpc.module.this_vpc.aws_route_table.intra[0]: Refreshing state... [id=rtb-06c01341966f74d89]
module.vpc.module.this_vpc.aws_subnet.database[2]: Refreshing state... [id=subnet-00447ac37f54755ac]
module.vpc.module.this_vpc.aws_subnet.public[2]: Refreshing state... [id=subnet-0a7e1bac565810db3]
module.vpc.module.this_vpc.aws_subnet.database[1]: Refreshing state... [id=subnet-04d1077df30eeef26]
module.vpc.module.this_vpc.aws_subnet.public[1]: Refreshing state... [id=subnet-0bd73d3aa3d69785a]
module.vpc.module.this_vpc.aws_subnet.public[0]: Refreshing state... [id=subnet-011c4f46c03588755]
module.vpc.module.this_vpc.aws_vpc_endpoint.dynamodb[0]: Refreshing state... [id=vpce-05413cc3d35e80482]
module.vpc.module.this_vpc.aws_subnet.intra[2]: Refreshing state... [id=subnet-08a77ff45e8530203]
module.vpc.module.this_vpc.aws_subnet.intra[0]: Refreshing state... [id=subnet-031fb972b887fa159]
module.vpc.module.this_vpc.aws_internet_gateway.this[0]: Refreshing state... [id=igw-041b2166c9d611b4a]
module.vpc.module.this_vpc.aws_subnet.intra[1]: Refreshing state... [id=subnet-00be6d1cb1ef273c3]
module.vpc.module.this_vpc.aws_vpc_endpoint.s3[0]: Refreshing state... [id=vpce-0a2934021a49578e7]
module.vpc.module.this_vpc.aws_subnet.private[1]: Refreshing state... [id=subnet-000709012842e43c6]
module.vpc.module.this_vpc.aws_subnet.private[2]: Refreshing state... [id=subnet-050529b642607e961]
module.vpc.module.this_vpc.aws_subnet.private[0]: Refreshing state... [id=subnet-0a4e0adf19819e68d]
module.vpc.module.this_vpc.aws_route_table.database[0]: Refreshing state... [id=rtb-0501d474b8b0c9234]
module.vpc.module.this_vpc.aws_route_table.public[0]: Refreshing state... [id=rtb-0d4717cb346f016df]
module.vpc.module.this_vpc.aws_route_table.private[0]: Refreshing state... [id=rtb-0b7d9ab8e2caf13c6]
module.vpc.module.this_vpc.aws_route_table.private[1]: Refreshing state... [id=rtb-0bf6ae73c1c08a42a]
module.vpc.module.this_vpc.aws_route_table.private[2]: Refreshing state... [id=rtb-094156867da30f108]
module.vpc.module.this_vpc.aws_egress_only_internet_gateway.this[0]: Refreshing state... [id=eigw-0419f8df14d9ef041]
module.vpc.module.this_vpc.aws_db_subnet_group.database[0]: Refreshing state... [id=my-vpc]
module.vpc.module.this_vpc.aws_nat_gateway.this[2]: Refreshing state... [id=nat-0af53719688cfd804]
module.vpc.module.this_vpc.aws_nat_gateway.this[1]: Refreshing state... [id=nat-00451d38c7fa6ef04]
module.vpc.module.this_vpc.aws_nat_gateway.this[0]: Refreshing state... [id=nat-07a248aa20d3ce6c2]
module.vpc.module.this_vpc.aws_route_table_association.intra[0]: Refreshing state... [id=rtbassoc-0450bde052c824cc0]
module.vpc.module.this_vpc.aws_route_table_association.intra[2]: Refreshing state... [id=rtbassoc-0788f98ce5c820075]
module.vpc.module.this_vpc.aws_route_table_association.intra[1]: Refreshing state... [id=rtbassoc-032db7dd47cc9a0ac]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.intra_dynamodb[0]: Refreshing state... [id=a-vpce-05413cc3d35e804823353412639]
module.vpc.module.this_vpc.aws_route_table_association.public[0]: Refreshing state... [id=rtbassoc-0660057fd826b7bc3]
module.vpc.module.this_vpc.aws_route_table_association.public[2]: Refreshing state... [id=rtbassoc-0ec6c16b2aa495265]
module.vpc.module.this_vpc.aws_route_table_association.public[1]: Refreshing state... [id=rtbassoc-06f7949755ed13927]
module.vpc.module.this_vpc.aws_route.public_internet_gateway_ipv6[0]: Refreshing state... [id=r-rtb-0d4717cb346f016df2750132062]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.public_dynamodb[0]: Refreshing state... [id=a-vpce-05413cc3d35e80482687498049]
module.vpc.module.this_vpc.aws_route.public_internet_gateway[0]: Refreshing state... [id=r-rtb-0d4717cb346f016df1080289494]
module.vpc.module.this_vpc.aws_route_table_association.private[1]: Refreshing state... [id=rtbassoc-0bee98503a933c584]
module.vpc.module.this_vpc.aws_route_table_association.private[2]: Refreshing state... [id=rtbassoc-08cd55715fcca0614]
module.vpc.module.this_vpc.aws_route_table_association.private[0]: Refreshing state... [id=rtbassoc-0106774c41c2aa98a]
module.vpc.module.this_vpc.aws_route_table_association.database[2]: Refreshing state... [id=rtbassoc-048809c0fe35c0ed5]
module.vpc.module.this_vpc.aws_route_table_association.database[0]: Refreshing state... [id=rtbassoc-0248fe206fd215821]
module.vpc.module.this_vpc.aws_route_table_association.database[1]: Refreshing state... [id=rtbassoc-02d66792ce6ffbda4]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.private_dynamodb[2]: Refreshing state... [id=a-vpce-05413cc3d35e80482306041323]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.private_dynamodb[1]: Refreshing state... [id=a-vpce-05413cc3d35e804822633274326]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.private_dynamodb[0]: Refreshing state... [id=a-vpce-05413cc3d35e804823925644090]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.intra_s3[0]: Refreshing state... [id=a-vpce-0a2934021a49578e73353412639]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.private_s3[2]: Refreshing state... [id=a-vpce-0a2934021a49578e7306041323]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.private_s3[0]: Refreshing state... [id=a-vpce-0a2934021a49578e73925644090]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.private_s3[1]: Refreshing state... [id=a-vpce-0a2934021a49578e72633274326]
module.vpc.module.this_vpc.aws_vpc_endpoint_route_table_association.public_s3[0]: Refreshing state... [id=a-vpce-0a2934021a49578e7687498049]
module.vpc.module.this_vpc.aws_route.private_ipv6_egress[2]: Refreshing state... [id=r-rtb-094156867da30f1082750132062]
module.vpc.module.this_vpc.aws_route.private_ipv6_egress[0]: Refreshing state... [id=r-rtb-0b7d9ab8e2caf13c62750132062]
module.vpc.module.this_vpc.aws_route.private_ipv6_egress[1]: Refreshing state... [id=r-rtb-0bf6ae73c1c08a42a2750132062]
module.vpc.module.this_vpc.aws_route.private_nat_gateway[0]: Refreshing state... [id=r-rtb-0b7d9ab8e2caf13c61080289494]
module.vpc.module.this_vpc.aws_route.private_nat_gateway[2]: Refreshing state... [id=r-rtb-094156867da30f1081080289494]
module.vpc.module.this_vpc.aws_route.private_nat_gateway[1]: Refreshing state... [id=r-rtb-0bf6ae73c1c08a42a1080289494]

------------------------------------------------------------------------

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.
```

</p>
</details>

Downloading the new v0.19.30 binary, removing the `.terraform` directory in the same module path, and re-running the `terragrunt plan` results in the following error:

```shell
$ rm -rf .terraform

$ ./terragrunt --version
terragrunt version v0.19.30

$ ./terragrunt plan

[terragrunt] 2019/10/08 10:22:12 Reading Terragrunt config file at /path/to/my/terragrunt/module/terragrunt.hcl
[terragrunt] [/path/to/my/terragrunt/module] 2019/10/08 10:22:12 Running command: terraform --version
[terragrunt] [/path/to/my/terragrunt/module] 2019/10/08 10:22:12 Initializing remote state for the s3 backend
[terragrunt] 2019/10/08 10:22:12 Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): AccessDenied: User: arn:aws:iam::111222333444:user/myawsuser is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::"111222333444":role/TerragruntAdministrator
	status code: 403, request id: baa64da5-e9a4-11e9-9cb6-bf91d2d562e4
[terragrunt] 2019/10/08 10:22:12 Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

With here the full debug output:

<details><summary><b>Terragrunt v0.19.30 attempted plan output</b></summary>
<p>

```shell
$ TERRAGRUNT_DEBUG='true' ./terragrunt plan

[terragrunt] 2019/10/08 10:24:21 Reading Terragrunt config file at /path/to/my/terragrunt/module/terragrunt.hcl
[terragrunt] [/path/to/my/terragrunt/module] 2019/10/08 10:24:22 Running command: terraform --version
[terragrunt] [/path/to/my/terragrunt/module] 2019/10/08 10:24:22 Initializing remote state for the s3 backend
[terragrunt] 2019/10/08 10:24:22 *awserr.requestError Error finding AWS credentials (did you set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environment variables?): AccessDenied: User: arn:aws:iam::111222333444:user/myawsuser is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::"111222333444":role/TerragruntAdministrator
	status code: 403, request id: 082292f8-e9a5-11e9-9395-6b92a3d12572
/go/src/github.com/gruntwork-io/terragrunt/aws_helper/config.go:73 (0x1782b47)
/go/src/github.com/gruntwork-io/terragrunt/remote/remote_state_s3.go:631 (0x17e5ee9)
/go/src/github.com/gruntwork-io/terragrunt/remote/remote_state_s3.go:205 (0x17e26e4)
<autogenerated>:1 (0x17e7a46)
/go/src/github.com/gruntwork-io/terragrunt/remote/remote_state.go:60 (0x17de262)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:497 (0x18fdc13)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:412 (0x18fd3e3)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:618 (0x18fe65a)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:561 (0x18fe33b)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:416 (0x18fd2c1)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:313 (0x18fc10a)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:227 (0x18fbc11)
/go/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:208 (0x18fb93e)
/go/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:490 (0x1899018)
/go/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:264 (0x189728d)
/go/src/github.com/gruntwork-io/terragrunt/main.go:23 (0x1904284)
/usr/local/go/src/runtime/proc.go:201 (0x102bba7)
/usr/local/go/src/runtime/asm_amd64.s:1333 (0x1057b91)

[terragrunt] 2019/10/08 10:24:22 Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

</p>
</details>

## Expected Behavior

Terragrunt v0.19.30 should successfully `plan` since nothing about the root `terragrunt.hcl` file was changed, nor was the module.

## Actual Behavior

Terragrunt v0.19.30 results in the error described above, where the AWS user is denied from assuming the role.

## Specifications

  - Version: Terragrunt v0.19.30 and Terraform v0.12.10
  - Platform: macOS 10.14.6
<br />
***


**mcalhoun** commented *Oct 8, 2019*

Thanks for reporting this @sc250024! It does looks like we're wrapping the output in extraneous quotes. I'll take a look at getting that fixed now and will update this issue when a new release is issued. 
***

**sc250024** commented *Oct 8, 2019*

@mcalhoun No problem, thanks for the quick response!
***

